### PR TITLE
Tag `GpuWindow` child expressions for GPU execution

### LINKED ISSUE #5984: [BUG] DATABRICKS: NullPointerException: format is null  in 22.08 (works fine with 22.06)
**Describe the bug**
Running the following scala code on databricks results in an NullPointerException. (See Environment details section)

Note: The same code works fine on local workstation 

**Steps/Code to reproduce bug**

```
import org.apache.hadoop.fs.Path
import org.apache.spark.sql.expressions.Window
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Column, DataFrame, SparkSession}

val rawSchema = StructType(Array(
      StructField("loan_id", LongType),
      StructField("orig_date", StringType),
      StructField("first_pay_date", StringType),
      StructField("monthly_reporting_period", StringType)
      )
    )

val sparkSession = SparkSession.builder.appName("nullpointBug").getOrCreate
val reader = sparkSession.read.option("header", false).schema(rawSchema)
val rawDf = reader.option("nullValue", "")
      .schema(rawSchema)
      .csv("/saralihalli/nullpointbug/csv/")

rawDf.select(col("loan_id")
      ,date_format(to_date(col("orig_date"),"MMyyyy"), "MM/yyyy").as("orig_date")
      ,date_format(to_date(col("first_pay_date"),"MMyyyy"), "MM/yyyy").as("first_pay_date")
      ,dense_rank().over(Window.partitionBy("loan_id").orderBy(to_date(col("monthly_reporting_period"),"MMyyyy"))).as("rank")
      ).show()
```

Use the following data to create a csv file
```
100007365142,121999,022000,022000
100007365142,121999,022000,032000
100007365142,121999,022000,042000
100007365142,121999,022000,052000
100007365142,121999,022000,062000
100007365142,121999,022000,072000
100007365142,121999,022000,082000
100007365142,121999,022000,092000
100007365142,121999,022000,102000
100007365142,121999,022000,112000
```

**Environment details (please complete the following information)**

Databricks: 10.4 LTS ML (includes Apache Spark 3.2.1, GPU, Scala 2.12)
Fails when 22.08 jar is used. However, the same scala code works when 22.06 jar is used

22.08 Jar: rapids-4-spark_2.12-22.08.0-20220710.113328-25.jar 

**Additional context**

```
rawDf.select(col("loan_id")
      ,dense_rank().over(Window.partitionBy("loan_id").orderBy(to_date(col("monthly_reporting_period"),"MMyyyy"))).as("rank")
      ).show()

```
and

```
rawDf.select(col("loan_id")
      ,date_format(to_date(col("orig_date"),"MMyyyy"), "MM/yyyy").as("orig_date")
      ,date_format(to_date(col("first_pay_date"),"MMyyyy"), "MM/yyyy").as("first_pay_date")
      ).show()
```

individually pass, but when combined results in error.

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #5984.

This change tags the child expressions of a GpuWindowExecMeta, to ensure
that all dependency expressions initialized correctly.
    
Certain child expressions might not be completely initialized even after
`GpuOverrides#wrapExpr()` has been called. For instance, `UnixTimeExprMeta`
postpones some of its initialization until `tagExprForGpu()` is called.
    
By declaring `windowExpressions`, `partitionSpec`, `orderSpec`, etc. as
`childExprs` of `GpuWindowExecMeta`, one guarantees that the
`GpuOverrides` infrastructure will tag them for GPU execution.

Signed-off-by: MithunR <mythrocks@gmail.com>

### CODE REVIEW

**Comment by mythrocks** [comment]
(Tentative fix. Still considering other options to sort out #5984.)

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 1]
```diff
@@ -374,19 +374,26 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
    rule: DataFromReplacementRule)
   extends BinaryExprMeta[A](expr, conf, parent, rule) {
 
-  var sparkFormat: String = _
-  var strfFormat: String = _
+  val optionalRightStringLit: Option[String] =
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:377]: Any reason for this not to be lazy, like the members below?

Comment (mythrocks) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:377]: Hmm. None that I can see. Good catch.

**Comment by mythrocks** [comment]
I'll post a better solution to this shortly.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 1 above)
Comment (mythrocks) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala]: Closing this in favour of a better solution.

**Comment by mythrocks** [comment]
> I'll post a better solution to this shortly.

The solution has been modified to have `GpuWindowExecMeta` get its child expressions tagged for GPU.
It was a mistake to consider that `GpuOverrides#wrapExpr()` produces fully constructed expressions; certain of them might not be fully initialized until after `tagExprForGpu()` is called.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Comment by mythrocks** [comment]
Thank you for reviewing, @jlowe. I've merged this now.

**Pull request merged**